### PR TITLE
License extraction opt-in

### PR DESCRIPTION
**Pull request by johbo**
We try to extract the license from the egg-info information of required packages. The use case is to use the nix dependency tree to generate an overview of all third party packages and also a semi-automatic license verification in case your corporate environment or project poses some constraints.

It's based on a best effort principle: If the information is there and we can understand it, then we try to set `meta.license` based on `pkgs.lib.licenses`, otherwise we take the string value.

In our project we use this to get the license information for about 90% of our dependencies, the rest is something we fill in manually with an override layer. That seems to be a good trade-off so far.

What do you think about adding support for this? The code is not yet fully ready for inclusion I think, but maybe good enough for some early feedback on it. If you think that's useful, we can bring it into good shape so that it is ready for inclusion.

### CODE REVIEW

**Comment by ktosiek** [comment]
Thanks, I like the idea!
For loading the list of licenses it would be enough to make it lazy/cached, I'll try to look over the whole diff soon.

File: pip2nix/models/package.py [hunk 1]
```diff
@@ -67,13 +127,109 @@ def to_nix(self):
...
+
+        # Convert license strings to nix.
+        nix_licenses = set()
+        for lic in licenses:
+            nix_licenses.add(license_to_nix(lic))
+
+        template = '[ {licenses} ]'
+        return template.format(licenses=' '.join(nix_licenses))
+
+    def get_licenses_from_setup(self):
```
Comment (ktosiek) [inline pip2nix/models/package.py]: Can this be done with egg_info writers (like it works with tests_require.txt already)? Then it wouldn't be necessary to rerun the setup.py here.

Edit: ok, looks like it's not actually used - can you remove that?

**Comment by ktosiek** [comment]
Making the `lib.licenses` loading lazy and removing the dead code should make this mergeable.
Later I'll probably have to move creating `PythonPackage`s to some factory that will know about licenses and settings, I don't really like having much logic inside the `PythonPackage` class.
Also, I think this could be enabled by default - if I understand correctly, worst case scenario is having some licenses duplicated/only having full_name, which is not too bad.

**Comment by johbo** [comment]
Thanks for the feedback, we'll soon work on bringing this into shape then :)

**Comment by marcinkuzminski** [comment]
@johbo could we update this and hopefully have it merged soon ?

**Comment by ergo** [comment]
bump

**Comment by johbo** [comment]
I'll do my best™

**Comment by johbo** [comment]
got back to it, changes are in https://github.com/johbo/pip2nix/tree/licenses, need a hand from @mbld to update the PR

**Comment by johbo** [comment]
Now I am happy that I added a small test :D Quite nice how `nix-build release.nix pip2nix.pythonXX` can be used to test in different environments 👍

**Comment by johbo** [comment]
Hmm, the python33 test run is failing, that's odd

**Comment by johbo** [comment]
Locally I got it working with a more recent version of py.test, hope that's also going to do the trick on travis.

**Comment by johbo** [comment]
Very odd, tried `nix-build '<nixpkgs>' -A python33Packages.pytest` locally. The interesting part: On the first run this fails with the issue I see in travis, but if I try it again it just works. Once I run a garbage collection, it fails again one time. Bit clueless about this.

I'll remove the dependency update, since this did not help.

**Comment by ktosiek** [comment]
It looks like it's a problem with build purity on 16.09 and unstable, the tests pass locally for me too.

**Review by ktosiek — APPROVED** [review]
This looks OK. If the dependencies bump was done by just running pip2nix itelf, please leave it in.

**Comment by johbo** [comment]
I'll bring it back in a second

**Pull request merged**